lib/stages/compressLostSamples.cpp

```diff
@@ -28,50 +27,57 @@ compressLostSamples::compressLostSamples(Config& config_, const std::string& uni
 
-    in_buf = get_buffer("lost_samples_buf");
+    in_buf = get_buffer("in_buf");
     register_consumer(in_buf, unique_name.c_str());
 
-    out_buf = get_buffer("compressed_lost_samples_buf");
+    out_buf = get_buffer("out_buf");
     register_producer(out_buf, unique_name.c_str());
+
+    if (_samples_per_data_set != (uint32_t)in_buf->frame_size) {
+        throw std::runtime_error("compressLostSamples in_frame has the wrong size.");
+    }
+
+    if ((_samples_per_data_set / _compression_factor)
+        != (uint32_t)out_buf->frame_size / sizeof(uint32_t)) {
+        throw std::runtime_error("compressLostSamples out_frame has the wrong size.");
+    }
 }
 
 compressLostSamples::~compressLostSamples() {}
 
 void compressLostSamples::main_thread() {
 
-    uint in_buffer_ID = 0;
+    frameID in_buffer_ID(in_buf);
     uint8_t* in_frame;
-    int out_buffer_ID = 0;
-
-    // Get the first output buffer which will always be id = 0 to start.
-    uint8_t* out_frame = wait_for_empty_frame(out_buf, unique_name.c_str(), out_buffer_ID);
-    if (out_frame == nullptr)
-        goto end_loop;
+    frameID out_buffer_ID(out_buf);
+    uint32_t* out_frame;
 
     while (!stop_thread) {
         // Get an input buffer, This call is blocking!
         in_frame = wait_for_full_frame(in_buf, unique_name.c_str(), in_buffer_ID);
         if (in_frame == nullptr)
-            goto end_loop;
+            break;
+
+        out_frame = (uint32_t*)wait_for_empty_frame(out_buf, unique_name.c_str(), out_buffer_ID);
+        if (out_frame == nullptr)
+            break;
 
-        // Information on dropped packets
-        uint8_t* lost_samples_frame = in_buf->frames[in_buffer_ID];
-        uint8_t* compressed_lost_samples_frame = out_buf->frames[out_buffer_ID];
         uint32_t total_lost_samples = 0;
 
         // Compress lost samples buffer by checking each sample for a flag
-        for (uint sample = 0; sample < _samples_per_data_set;
-             sample += NUM_SETS_OF_SUB_FREQS * _factor_upchan) {
-            compressed_lost_samples_frame[sample / (NUM_SETS_OF_SUB_FREQS * _factor_upchan)] = 0;
-            for (uint freq = 0; freq < NUM_SETS_OF_SUB_FREQS * _factor_upchan; freq++) {
-                if (lost_samples_frame[sample + freq]) {
-                    compressed_lost_samples_frame[sample
-                                                  / (NUM_SETS_OF_SUB_FREQS * _factor_upchan)] = 1;
-                    total_lost_samples += NUM_SETS_OF_SUB_FREQS * _factor_upchan;
-                    break;
-                }
+        for (uint32_t sample = 0; sample < _samples_per_data_set; sample += _compression_factor) {
+            // assert(sample/_compression_factor < (uint32_t)out_buf->frame_size);
+            out_frame[sample / _compression_factor] = 0;
+            for (uint32_t sub_index = 0; sub_index < _compression_factor; sub_index++) {
+                out_frame[sample / _compression_factor] += in_frame[sample + sub_index];
+            }
```

Comment:
Why not have:
```
               if(_zero_all_in_group && in_frame[sample + sub_index]) {
                    out_frame[sample / _compression_factor] = 1;
                    total_lost_samples += _compression_factor;
                    break;
                }
                else {
                    out_frame[sample / _compression_factor] += in_frame[sample + sub_index];
                }
            }
            if (!_zero_all_in_group) {                                                                                                                                                                               
                total_lost_samples += out_frame[sample / _compression_factor];
            }
```
which would mean you could exit the loop early if you find any non-zero values in the lost samples buffer. Slight optimisation but it might not matter as it introduces more `if` statements.

Reply:
Yeah that seems like a reasonable optimization.